util/lock.go

```diff
@@ -0,0 +1,170 @@
+package util
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"github.com/timescale/prometheus-postgresql-adapter/log"
+	"strconv"
+	"sync"
+	"time"
+)
+
+const (
+	connCheckInterval = time.Millisecond * 250
+)
+
+// PgAdvisoryLock is implementation of leader election protocol based on PostgreSQL advisory locks. All adapters withing a HA group are trying
+// to obtain an advisory lock for particular group. The one who holds the lock can write to the database. Due to the fact
+// that Prometheus HA setup provides no consistency guarantees this implementation is best effort in regards
+// to metrics that is written (some duplicates or data loss are possible during failover)
+type PgAdvisoryLock struct {
+	conn     *sql.Conn
+	mutex    sync.RWMutex
+	obtained bool
+	connPool *sql.DB
+
+	quit        chan bool
+	ticker      *time.Ticker
+	groupLockId int
+	closed      bool
+}
+
+func NewPgAdvisoryLock(groupLockId int, connPool *sql.DB) (*PgAdvisoryLock, error) {
+	lockConn, err := getConn(connPool)
+	if err != nil {
+		return nil, err
+	}
+	if err := checkConnection(lockConn); err != nil {
+		return nil, err
+	}
+	lock := &PgAdvisoryLock{conn: lockConn, obtained: false, ticker: time.NewTicker(connCheckInterval), quit: make(chan bool), groupLockId: groupLockId, closed: false}
+	_, err = lock.TryLock()
+	if err != nil {
+		return nil, err
+	}
+	go lock.connCheck()
+	return lock, nil
+}
+
+func getConn(pool *sql.DB) (*sql.Conn, error) {
+	lockConn, err := pool.Conn(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("error getting DB connection: %v", err)
+	}
+	return lockConn, nil
+}
+
+func (l *PgAdvisoryLock) Id() string {
+	return strconv.Itoa(l.groupLockId)
+}
+
+func (l *PgAdvisoryLock) BecomeLeader() (bool, error) {
+	return l.TryLock()
+}
+
+func (l *PgAdvisoryLock) IsLeader() (bool, error) {
+	return l.Locked(), nil
+}
+
+func (l *PgAdvisoryLock) Resign() error {
+	return l.Release()
+}
+
+func (l *PgAdvisoryLock) TryLock() (bool, error) {
+	l.mutex.Lock()
+	if l.closed {
+		return false, fmt.Errorf("lock is closed. please create a new lock")
+	}
+	defer l.mutex.Unlock()
+	if l.obtained {
+		return l.obtained, nil
+	}
+	err := checkConnection(l.conn)
+	if err != nil {
+		conn, err := getConn(l.connPool)
+		if err != nil {
+			return false, err
+		}
+		l.conn = conn
+	}
+	rows, err := l.conn.QueryContext(context.Background(), fmt.Sprintf("SELECT pg_try_advisory_lock(%d)", l.groupLockId))
```

Comment:
Using sprintf for creating sql statements should be avoided. You should use something like the following:
```
QueryContext(context.Background(),"SELECT pg_try_advisory_lock(?)", l.groupLockId)
```

Reply:
For some reason `pq` driver does not support that syntax when invoking SQL functions. I tried even `SELECT pg_try_advisory_lock(key => $1)` but no luck

Reply:
I think correct syntax is ```SELECT pg_try_advisory_lock($1)``` which works for me